src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+
+  constructor(options: DirectLineOptions, parentDirectLine: DirectLine) {
+    this.token = options.secret || options.token;
+
+    if (options.token) {
+      this.refreshToken();
+    }
+
+    if (options.domain) {
+      this.domain = options.domain;
+    }
+
+    if (options.conversationId) {
+      this.conversationId = options.conversationId;
+    }
+
+    this.parentDirectLine = parentDirectLine;
+
+    this._botAgent = this.getBotAgent(options.botAgent);
+
+    this.activity$ = this.streamingWebSocketActivity$().share();
+      this.parentDirectLine.activity$ = this.activity$;
+      this.parentDirectLine.connectionStatus$ = this.connectionStatus$;
+      this.parentDirectLine.postActivity = this.postActivity.bind(this);
+      this.parentDirectLine.end = this.end.bind(this);
+  }
+
+  public reconnect(conversation: Conversation) {
+    this.conversationId = conversation.conversationId;
+    this.token = conversation.token;
+    this.connectAsync().then(_ => _);
+  }
+
+  end() {
+    this.connectionStatus$.next(ConnectionStatus.Ended);
+    this.streamConnection.disconnect();
+  }
+
+  private commonHeaders() {
+    return {
+      "Authorization": `Bearer ${this.token}`,
+      "x-ms-bot-agent": this._botAgent
+    };
+  }
+
+  private getBotAgent(customAgent: string = ''): string {
+    let clientAgent = 'directlinejs'
+
+    if (customAgent) {
+      clientAgent += `; ${customAgent}`
+    }
+
+    return `${DIRECT_LINE_VERSION} (${clientAgent})`;
+  }
+
+  private async refreshToken(firstCall = true, retryCount = 0) {
+    if (firstCall) {
+      setTimeout(async () => await this.refreshToken(false, 0), refreshTokenInterval);
+      return;
+    }
+
+    if (this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    await this.waitUntilOnline();
+
+    Observable.ajax({
+      method: "POST",
+      url: `${this.domain}/tokens/refresh`,
+      timeout,
+      headers: {
+        ...this.commonHeaders()
+      }
+    })
+      .subscribe(
+        (ajaxResponse) => {
+          this.token = ajaxResponse.response.token as string;
+          setTimeout(() => this.refreshToken(false, MAX_RETRY_COUNT), refreshTokenInterval);
+        },
+        (e) => {
+          if (e.status === 403 || e.status === 404) {
+            console.warn("Error refreshing token " + e.status);
+          }else{
+            if (retryCount > 0) {
+              console.warn("Error refreshing token " + e.status + " " + retryCount + " retries left");
+              this.refreshToken(false, retryCount-1);
+            }else{
+              this.token = null;
+              console.warn("Error refreshing token " + e.status + " Retries exhausted");
+            }
+          }
+        })
+  }
+
+  postActivity(activity: Activity) {
+    if (activity.type === "message" && activity.attachments && activity.attachments.length > 0)
+      return this.postMessageWithAttachments(activity);
+
+    let resp$ = Observable.create(subscriber => {
```

Comment:
Use `async` and simplify the code below.

```suggestion
    return Observable.create(async subscriber => {
```